table_engine/src/partition/rule/key.rs

```diff
@@ -0,0 +1,346 @@
+        // for example:
+        //      key_col1: [f1, f2, f3]
+        //      key_col2: [f4, f5]
+        // will convert to:
+        //      group1: [key_col1: f1, key_col2: f4]
+        //      group2: [key_col1: f1, key_col2: f5]
+        //      group3: [key_col1: f2, key_col2: f4]
+        //      ...
+        let empty_filter = filter_by_columns.iter().find(|filter| filter.is_empty());
+        if empty_filter.is_some() {
+            return Ok(Vec::default());
+        }
+
+        let groups = filter_by_columns
+            .into_iter()
+            .map(|filters| filters.into_iter())
+            .multi_cartesian_product()
+            .collect_vec();
+        debug!(
+            "KeyRule get candidate partition keys groups, groups:{:?}",
+            groups
+        );
+
+        Ok(groups)
+    }
+
+    fn compute_partition_for_inserted_row(
+        &self,
+        row: &Row,
+        target_column_idxs: &[usize],
+        buf: &mut BytesMut,
+    ) -> usize {
+        let partition_keys = target_column_idxs
+            .iter()
+            .map(|col_idx| &row[*col_idx])
+            .collect_vec();
+        compute_partition(&partition_keys, self.partition_num, buf)
+    }
+}
+
+impl PartitionRule for KeyRule {
+    fn columns(&self) -> Vec<String> {
+        self.typed_columns
+            .iter()
+            .map(|typed_col| typed_col.column.clone())
+            .collect()
+    }
+
+    fn locate_partitions_for_write(&self, row_group: &RowGroup) -> Result<Vec<usize>> {
+        // Extract idxs.
+        let typed_idxs = self
+            .typed_columns
+            .iter()
+            .map(|typed_col| row_group.schema().index_of(typed_col.column.as_str()))
+            .collect::<Option<Vec<_>>>()
+            .with_context(|| LocateWritePartition {
+                msg: format!(
+                    "in key partition not all key columns found in schema, key columns:{:?}",
+                    self.typed_columns
+                ),
+            })?;
+
+        // Compute partitions.
+        let mut buf = BytesMut::new();
+        let partitions = row_group
+            .iter()
+            .map(|row| self.compute_partition_for_inserted_row(row, &typed_idxs, &mut buf))
+            .collect_vec();
+        Ok(partitions)
+    }
+
+    fn locate_partitions_for_read(&self, filters: &[PartitionFilter]) -> Result<Vec<usize>> {
+        let all_partitions = (0..self.partition_num as usize).into_iter().collect_vec();
+
+        // Filters are empty.
+        if filters.is_empty() {
+            return Ok(all_partitions);
+        }
+
+        // Group the filters by their columns.
+        let candidate_partition_keys_groups = self.get_candidate_partition_keys_groups(filters)?;
+        if candidate_partition_keys_groups.is_empty() {
+            return Ok(all_partitions);
+        }
+
+        let mut partitions = HashSet::new();
+        let mut buf = BytesMut::new();
+        for group in candidate_partition_keys_groups {
```

Comment:
Compute partitions in different groups, and then find the intersection of these values.